imap_processing/ultra/l1b/de.py

```diff
@@ -23,36 +39,161 @@ def calculate_de(de_dataset: xr.Dataset, name: str) -> xr.Dataset:
+    ph_energy = get_energy_pulse_height(
+        de_dataset["STOP_TYPE"].data[ph_indices],
+        de_dataset["ENERGY_PH"].data[ph_indices],
+        ph_xb,
+        ph_yb,
+    )
+    ph_r = get_path_length((xf[ph_indices], ph_yf), (ph_xb, ph_yb), ph_d)
+    ph_bin = determine_species_pulse_height(ph_energy, ph_tof, ph_r)
+    ph_etof, ph_xc = get_coincidence_positions(
+        de_dataset.isel(epoch=ph_indices), ph_t2, f"ultra{sensor}"
+    )
+    ph_ctof = get_ctof(ph_tof, ph_r, "PH")
+
+    # SSD
+    ssd_indices = np.nonzero(np.isin(de_dataset["STOP_TYPE"], StopType.SSD.value))[0]
+    ssd_tof = get_ssd_tof(de_dataset, xf)
+    ssd_yb, ssd_tof_offset, ssd_number = get_ssd_back_position_and_tof_offset(
+        de_dataset
+    )
+    ssd_xb = np.zeros(len(ssd_yb))
+    ssd_d, ssd_yf = get_front_y_position(
+        de_dataset["START_TYPE"].data[ssd_indices], ssd_yb
+    )
+    ssd_energy = get_energy_ssd(de_dataset, ssd_number)
+    ssd_r = get_path_length((xf[ssd_indices], ssd_yf), (ssd_xb, ssd_yb), ssd_d)
+    ssd_bin = determine_species_ssd(
+        ssd_energy,
+        ssd_tof,
+        ssd_r,
+    )
+    ssd_ctof = get_ctof(ssd_tof, ssd_r, "SSD")
+
+    # Combine ph_yb and ssd_yb along with their indices
+    combined_indices = np.argsort(np.concatenate((ph_indices, ssd_indices)))
+    de_dict["x_front"] = xf
+    yb = np.concatenate((ph_yb, ssd_yb))
+    de_dict["y_back"] = yb[combined_indices]
+    xb = np.concatenate((ph_xb, ssd_xb))
+    de_dict["x_back"] = xb[combined_indices]
+    xcoin = np.concatenate((ph_xc, np.zeros(len(ssd_indices))))
+    de_dict["x_coin"] = xcoin[combined_indices]
+    yf = np.concatenate((ph_yf, ssd_yf))
+    de_dict["y_front"] = yf[combined_indices]
+    d = np.concatenate((ph_d, ssd_d))
+    de_dict["front_back_distance"] = d[combined_indices]
+    r = np.concatenate((ph_r, ssd_r))
+    de_dict["path_length"] = r[combined_indices]
+    tof = np.concatenate((ph_tof, ssd_tof))
+    de_dict["tof_start_stop"] = tof[combined_indices]
+    etof = np.concatenate((ph_etof, np.zeros(len(ssd_indices))))
+    de_dict["tof_stop_coin"] = etof[combined_indices]
+
+    ctof = np.concatenate((ph_ctof, ssd_ctof))
+    de_dict["tof_corrected"] = ctof[combined_indices]
+
+    keys = [
+        "coincidence_type",
+        "start_type",
+        "event_type",
+        "de_event_met",
+    ]
+    dataset_keys = ["COIN_TYPE", "START_TYPE", "STOP_TYPE", "SHCOARSE"]
+
+    de_dict.update(
+        {key: de_dataset[dataset_key] for key, dataset_key in zip(keys, dataset_keys)}
+    )
+
+    vx_ultra, vy_ultra, vz_ultra = get_particle_velocity(
+        (de_dict["x_front"], de_dict["y_front"]),
+        (de_dict["x_back"], de_dict["y_back"]),
+        de_dict["front_back_distance"],
+        de_dict["tof_start_stop"],
+    )
+
+    # We need to fill velocities that have negative tof values.
+    de_dict["vx_ultra"] = np.full_like(
+        vx_ultra, np.finfo(np.float64).min, dtype=np.float64
```

Comment:
Is `np.finfo(np.float64).min` the same as `np.nan`? Might be worth checking. `np.nan` is much preferred IMO.